community/kernel/src/main/java/org/neo4j/kernel/impl/core/PathProxy.java

```diff
@@ -0,0 +1,284 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.core;
+
+import java.util.Arrays;
+import java.util.Iterator;
+
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.Path;
+import org.neo4j.graphdb.PropertyContainer;
+import org.neo4j.graphdb.Relationship;
+import org.neo4j.graphdb.traversal.Paths;
+
+import static org.neo4j.helpers.collection.Iterators.iteratorsEqual;
+
+public class PathProxy implements Path
+{
+    private final EmbeddedProxySPI proxySPI;
+    private final long[] nodes;
+    private final long[] relationships;
+    private final int[] directedTypes;
+
+    /**
+     * @param proxySPI
+     *         the API into the kernel.
+     * @param nodes
+     *         the ids of the nodes in the path, in order.
+     * @param relationships
+     *         the ids of the relationships in the path, in order.
+     * @param directedTypes
+     *         an encoding of the types and directions of the relationships.
+     *         An entry at position {@code i} of this array should be {@code typeId} if the relationship at {@code i}
+     *         has its start node at {@code i} and its end node at {@code i + 1}, and should be {@code ~typeId} if the
+     *         relationship at {@code i} has its start node at {@code i + 1} and its end node at {@code i}.
+     */
+    public PathProxy( EmbeddedProxySPI proxySPI, long[] nodes, long[] relationships, int[] directedTypes )
+    {
+        assert nodes.length == relationships.length + 1;
+        assert relationships.length == directedTypes.length;
+        this.proxySPI = proxySPI;
+        this.nodes = nodes;
+        this.relationships = relationships;
+        this.directedTypes = directedTypes;
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder string = new StringBuilder();
+        string.append( '(' ).append( nodes[0] ).append( ')' );
+        boolean inTx = true;
+        for ( int i = 0; i < relationships.length; i++ )
+        {
+            int type = directedTypes[i];
+            string.append( type < 0 ? "<-[" : "-[" );
+            string.append( relationships[i] );
+            if ( inTx )
+            {
+                try
+                {
+                    String name = proxySPI.getRelationshipTypeById( type < 0 ? ~type : type ).name();
+                    string.append( ':' ).append( name );
+                }
+                catch ( Exception e )
+                {
+                    inTx = false;
```

Comment:
Is there another way to figure out, if we are in a trnasaction currently?

Reply:
Through quite a bit of refactoring.

Reply:
There is a bit of redemption though, which makes this solution not-so-bad, and makes that refactoring not-really-worth-it, and that is that typically the `getRelationshipTypeById(...)`-method works even outside of a transaction - although it is not guaranteed to... (and would definitely fail if the database has been shut down)